stack/rest/src/main/java/org/apache/usergrid/rest/management/ManagementResource.java

```diff
@@ -441,6 +451,181 @@ public Viewable handleAuthorizeForm( @Context UriInfo ui, @FormParam( "response_
+
+
+    /**
+     * <p>
+     * Validates access token from other or "external" Usergrid system.
+     * Calls other system's /management/me endpoint to get the User associated with the access token.
+     * If user does not exist locally, then user and organization with the same name of user is created.
+     * If no user is returned from the other cluster, then this endpoint will return 401.
+     * </p>
+     *
+     * <p> Part of Usergrid Central SSO feature.
+     * See <a href="https://issues.apache.org/jira/browse/USERGRID-567">USERGRID-567</a>
+     * for details about Usergrid Central SSO.
+     * </p>
+     *
+     * @param ui             Information about calling URI.
+     * @param extAccessToken Access token from external Usergrid system.
+     * @param ttl            Time to live for token.
+     * @param callback       For JSONP support.
+     * @return               Returns JSON object with access_token field.
+     * @throws Exception     Returns 401 if access token cannot be validated
+     */
+    @GET
+    @Path( "/externaltoken" )
+    public Response validateExternalToken(
+                                @Context UriInfo ui,
+                                @QueryParam( "ext_access_token" ) String extAccessToken,
+                                @QueryParam( "ttl" ) @DefaultValue("-1") long ttl,
+                                @QueryParam( "callback" ) @DefaultValue( "" ) String callback )
+            throws Exception {
+
+
+        if ( StringUtils.isEmpty( properties.getProperty( USERGRID_CENTRAL_URL ))) {
+            throw new NotImplementedException( "External Token Validation Service is not configured" );
+        }
+
+        if ( extAccessToken == null ) {
+            throw new IllegalArgumentException("ext_access_token must be specified");
+        }
+
+        if ( ttl == -1 ) {
+            throw new IllegalArgumentException("ttl must be specified");
+        }
+
+        // look up user via UG Central's /management/me endpoint.
+
+        JsonNode accessInfoNode = getMeFromUgCentral( extAccessToken );
+
+        JsonNode userNode = accessInfoNode.get( "user" );
+        String username = userNode.get( "username" ).getTextValue();
+        String name     = userNode.get( "name" ).getTextValue();
+        String email    = userNode.get( "email" ).getTextValue();
+
+        // set dummy password to random string that nobody can guess, in SSO setup
+        // admin users should never be able to login directly to this Usergrid system
+        String dummyPassword = RandomStringUtils.randomAlphanumeric( 40 );
+
+        // if user does not exist locally then we need to fix that
+
+        final UUID userId;
+        final OrganizationInfo organizationInfo = management.getOrganizationByName(username);
+
+        if ( organizationInfo == null ) {
+
+            // create local user and personal organization, activate user.
+
+            OrganizationOwnerInfo ownerOrgInfo = management.createOwnerAndOrganization(
```

Comment:
Instead of creating an organization with the same name as the username, we should get the list of organizations from the accessInfoNode and create each of those listed.